src/plugins/workflow_designer/src/library/ReadAnnotationsWorker.cpp

```diff
@@ -242,30 +281,34 @@ void ReadAnnotationsTask::run() {
     m[BaseSlots::DATASET_SLOT().getId()] = datasetName;
 
-    QList<SharedAnnotationData> dataList;
-
-    for (GObject *go : qAsConst(annsObjList)) {
-        auto annsObj = dynamic_cast<AnnotationTableObject *>(go);
-        CHECK_EXT(annsObj != nullptr, stateInfo.setError("NULL annotations object"), );
-
-        if (!mergeAnnotations || annsObjList.size() == 1) {
-            doc->removeObject(go, DocumentObjectRemovalMode_Release);
-            const SharedDbiDataHandler tableId = context->getDataStorage()->putAnnotationTable(annsObj);
-            delete go;
+    // If "SPLIT" -- transfer (write to the result) tables of annotations from files as is.
+    // If "MERGE" and there is only one annotation table, transfer the table as is.
+    // If "MERGE_FILES", transfer the file tables as is and merge them into ReadAnnotationsWorker::sl_datasetEnded().
+    // Otherwise ("MERGE" with several tables of annotations in one file), merge these tables into one.
+    if (mergeAnnotations != ReadAnnotationsProto::MERGE || annsObjList.size() == 1) {
+        for (GObject *gObject : qAsConst(annsObjList)) {
+            auto annotationTable = dynamic_cast<AnnotationTableObject *>(gObject);
+            CHECK_EXT(annotationTable != nullptr, setError(L10N::nullPointerError("annotationTable")), );
+
+            QScopedPointer<AnnotationTableObject> annotationTableClone(qobject_cast<AnnotationTableObject *>(annotationTable->clone(dbiRef, stateInfo)));
+            CHECK_OP(stateInfo, )
+            CHECK_EXT(annotationTableClone != nullptr, setError(L10N::nullPointerError(QString("annotationTableClone '%1'").arg(gObject->getGObjectName()))), );
+            SharedDbiDataHandler tableId = context->getDataStorage()->putAnnotationTable(annotationTableClone.data());
             m[BaseSlots::ANNOTATION_TABLE_SLOT().getId()] = qVariantFromValue<SharedDbiDataHandler>(tableId);
             results.append(m);
-        } else {
-            foreach (Annotation *a, annsObj->getAnnotations()) {
-                dataList << a->getData();
-            }
         }
+        return;
     }
-
-    if (mergeAnnotations && annsObjList.size() > 1) {
-        const SharedDbiDataHandler tableId = context->getDataStorage()->putAnnotationTable(dataList, mergedAnnTableName);
-        m[BaseSlots::ANNOTATION_TABLE_SLOT().getId()] = qVariantFromValue<SharedDbiDataHandler>(tableId);
-        results.append(m);
+    QList<AnnotationTableObject *> annotationTables;
+    for (GObject *gObject : qAsConst(annsObjList)) {
+        auto annotationTable = dynamic_cast<AnnotationTableObject *>(gObject);
+        CHECK_EXT(annotationTable != nullptr, setError(L10N::nullPointerError("annotationTable")), );
+        annotationTables += annotationTable;
     }
+    std::unique_ptr<AnnotationTableObject> mergedTable = mergeAnnotationTables(dbiRef, annotationTables, mergedAnnTableName);
+    SharedDbiDataHandler tableId = context->getDataStorage()->putAnnotationTable(mergedTable.get());
```

Comment:
Here you need to delete all original annotation tables

Reply:
It will be removed with the document.